### ISSUE TITLE
subscriptions are lost during HMR

### ISSUE THREAD

**Issue by duoyue** [posva/pinia#991]
I have some code like this , it could output correctly when  app first loaded   or refresh ，but don‘t work after HMR. 

```javascript
const pinia = createPinia();
pinia.use(({ store }) => {
  store.$subscribe((mutations, state) => {
    console.log('trigger subscritions ', mutations, state);
  });
});
```

I have add some breakpoints in Pinia.js ，and found that `subscriptions ` is emty [] after HMR.
```javascript

// the  "subscriptions" is empty [] after HMR
function triggerSubscriptions(subscriptions, ...args) {
    subscriptions.forEach((callback) => {
        callback(...args);
    });
}

```

**Comment by posva** [comment]
Maybe this can be fixed but it does require some extra logic because any subscription that happens inside a plugin should not be transferred. There could be other cases of subscriptions that should not be transferred to avoid duplication.

Or maybe there is even a better solution within `acceptHMRUpdate`

**Comment by davidmyersdev** [comment]
@posva what do you mean by "any subscription that happens inside a plugin should not be transferred" here? I'm also running into this problem when trying to develop a plugin locally. The plugin only receives the `$subscribe` callback once before it is removed. Hot-reloading for plugins seems difficult since the initial plugin registration can have side effects on the current store state (e.g. hydrating the state from localStorage).

**Comment by davidmyersdev** [comment]
Additionally, it looks like this problem goes away when using `detached: true` by preventing the [`removeSubscription`](https://github.com/vuejs/pinia/blob/ec7caa2d49a0ece2e1ed250aea2749dfdd6dd3b8/packages/pinia/src/subscriptions.ts#L22) callback from firing on unmount. Is that configuration expected for plugins that need to subscribe to mutations/actions?

**Comment by Alanscut** [comment]
Passing `{ detached: true }` as the second parameter seems to fix this.

**Comment by emrysctl** [comment]
thanks

**Comment by zhu-hong** [comment]
When developing, pinia's hmr often doesn't work, which makes me very sick, either vite or vue cli

**Comment by posva** [comment]
> Additionally, it looks like this problem goes away when using `detached: true` by preventing the [`removeSubscription`](https://github.com/vuejs/pinia/blob/ec7caa2d49a0ece2e1ed250aea2749dfdd6dd3b8/packages/pinia/src/subscriptions.ts#L22) callback from firing on unmount. Is that configuration expected for plugins that need to subscribe to mutations/actions?

yes: in most cases you want to use `detached: true` in plugins so the subscription is never removed

**Comment by F0rsaken** [comment]
I think it would be very useful to mention it in `plugins` section of docs, especially in the  [`$subscribe` section](https://pinia.vuejs.org/core-concepts/plugins.html#calling-subscribe-inside-plugins) as if you develop a plugin and use `$subscribe` or `$onAction`, you almost certainly want to use `detach: true`

**Comment by posva** [comment]
@F0rsaken Totally! Could you do a PR?

**Comment by F0rsaken** [comment]
@posva sure, created PR with docs update https://github.com/vuejs/pinia/pull/1781 👍
